### PR TITLE
Add the SineReLU activation function as an advanced activation.

### PR DESCRIPTION
**Pull request by wilderrodrigues**
The Rectified Linear Unit with Sigmoid has been under development for almost a year now. It has been part of the IBM Watson AI XPRIZE Competition and benchmarked against the ReLU and Leaky ReLU, and still showing better results.

The following data sets were used:

* MNIST
* IMDB
* Kaggle Toxicity

The function will be demonstrated publicly at the Codemotion Amsterdam 2018 Conference and the DevDays in Vilnius, Lithuania. Both conferences are happening in May this year.

Jupyter notebooks using the function are available here: https://github.com/ekholabs/DLinK

Below, some benchmarking against ReLU and Leaky ReLU using Deep Nets and CNNs.

* Deep Net

image

* CNN

image

The function has also showed better results with DQN for the CartPole.v0 and ResNet 50.

Since I'm not linked to any academic institution, it has been proven difficult to make the function open. However, as a follower of the open source movement, I think it should be made available for others to test / use. I hope the Keras community can find a good use for it.

### CODE REVIEW

**Comment by wilderrodrigues** [comment]
Hi there @ahundt,

Would you have some time to take a look at this PR?

Thanks in advance and have a nice weekend.

Cheers,
Wilder

**Comment by ahundt** [comment]
Sure this can be merged here, I appreciate the advice on parameter choices in the comments. Two small things, can you add a unit test and a small example? 

The example can just be a modified version of one of the existing examples that uses the new activation and the unit test can be based off of something already here or in keras.

**Comment by wilderrodrigues** [comment]
Hi @ahundt,

Sure, will do that.

I use this activation function constantly and have plenty of examples with different architectures and datasets.

Will prepare them and update the PR soon.

**Comment by wilderrodrigues** [comment]
Weird TensorFlow installation error. Will force push and get the build running again:

```
The command "pip install tensorflow" failed and exited with 2 during .
```

**Comment by wilderrodrigues** [comment]
Added a test and documentation with examples on how to use the function with Deep Nets and CNNs, @ahundt.

Don't hesitate to let me know if that's okay or if you need something else.

Thanks again.

Cheers,
Wilder

File: keras_contrib/layers/advanced_activations.py [hunk 1]
```diff
@@ -283,3 +283,108 @@ def get_config(self):
...
+
+    # Input shape
+        Arbitrary. Use the keyword argument `input_shape`
+        (tuple of integers, does not include the samples axis)
+        when using this layer as the first layer in a model.
+
+    # Output shape
+        Same shape as the input.
+
+    # Usage
```
Comment (tboquet) [inline keras_contrib/layers/advanced_activations.py]: Can you put the examples at the end following this syntax:
https://github.com/keras-team/keras/blob/master/keras/layers/convolutional.py#L2245-L2256
One example should be enough.

Comment (wilderrodrigues) [reply keras_contrib/layers/advanced_activations.py]: Sure, will do that.

**Comment by wilderrodrigues** [comment]
Hi @tboquet ,

The example has been changed according to the best practices.

Thanks for the review. I hope that after the conferences where I'm going to publicly open the function it gets some attraction.

Cheers,
Wilder

**Comment by wilderrodrigues** [comment]
On an additional note, the equation will make the function look like this:

image

The epsilon hyper-parameter is used to control amplitude, length and oscillations.

**Comment by wilderrodrigues** [comment]
Hi @ahundt and @tboquet,

Any news on this PR?

Thanks again.

Cheers,
Wilder

**Comment by tboquet** [comment]
Can you report the performances of this activation on the datasets you mentioned (please include standard deviation)?
Can you also explain what is the difference between the periodic signal depending on the output (Z) and any structured noise that you would add (replace `m` by gaussian noise for example)?

File: keras_contrib/layers/advanced_activations.py [hunk 2]
```diff
@@ -283,3 +283,75 @@ def get_config(self):
         return dict(list(base_config.items()) + list(config.items()))
 
 get_custom_objects().update({'Swish': Swish})
+
+
+class ReLUs(Layer):
+    """Rectified Linear Unit with Sigmoid to generate oscilations.
```
Comment (tboquet) [inline keras_contrib/layers/advanced_activations.py]: Can you report the performances of this activation on the datasets you mentioned (please include standard deviation)?

File: keras_contrib/layers/advanced_activations.py [hunk 3]
```diff
@@ -283,3 +283,75 @@ def get_config(self):
...
+        ```
+    """
+
+    def __init__(self, epsilon=0.0055, **kwargs):
+        super(ReLUs, self).__init__(**kwargs)
+        self.supports_masking = True
+        self.epsilon = K.cast_to_floatx(epsilon)
+
+    def call(self, Z):
+        pi = K.variable((np.pi))
```
Comment (titu1994) [inline keras_contrib/layers/advanced_activations.py]: This should be precomputed and stored in an instance variable. Also, since it is a constant, there is no need to assign it to a tensor variable. 

It would be something like self.scale = np.exp(np.sqrt(np.pi)) inside build() and multiply inside call.

**Comment by titu1994** [comment]
In addition, the name ReLUs is highly confusing. It hides the fact that this is ReLU with sinusodial oscillations in it's negative range. 

If possible, can this be named to something that properly distinguishes it from standard ReLU? Since I see your paper has this nomenclature, you could provide an alias at the end of the file to still access it with "ReLUs".

**Comment by wilderrodrigues** [comment]
Thanks @tboquet and @titu1994 for the feedback. I will apply the changes suggested.

And @titu1994, I'm going for SineReLU. I really liked the name. :)

Cheers,
Wilder

**Comment by ahundt** [comment]
I agree with the comment @titu1994 made

**Comment by wilderrodrigues** [comment]
Got stuck with some work, guys. Also, have to retrained my models in order to get the STD for all datasets. I hope I will be done with the changes by the end of the week.

Cheers,
Wilder

File: keras_contrib/layers/advanced_activations.py (hunk 3 above)
Comment (wilderrodrigues) [reply keras_contrib/layers/advanced_activations.py]: Hi @titu1994,

Just started looking at the code changes before I got trining everything again.

I agree with the changes, since in this way some of the computations would be done in the __init__/build method and thus avoiding repetition (making things faster).

However, concerning "pi", it need to be a tensor in order to be used by K.exp(K.sqrt(...)). If that's not a tensor, but just a primitive type, we get the following:

```
'float' object has no attribute 'dtype'
```

I'm applying the changes and running tests to make sure all works fine before I push again.

Comment (titu1994) [reply keras_contrib/layers/advanced_activations.py]: Yes, that's why I suggested doing all constant operations in Numpy using no.exp(np.sqrt(np.pi)) * epsilon. 

Multiplying a tensor with a numpy constant is alright. Operating tensor ops on numpy values is not.

Comment (wilderrodrigues) [reply keras_contrib/layers/advanced_activations.py]: Ah, got it. I did not pay attention to the syntax and missed the "np" there. :)

Thanks!

File: keras_contrib/layers/advanced_activations.py (hunk 2 above)
Comment (wilderrodrigues) [reply keras_contrib/layers/advanced_activations.py]: Hi @tboquet,

Busy doing that now, but training the three datasets take a while. I also want to get a nice outcome for the SineRelu, so checking different epsilons here.

Will share the final loss, accuracy and standard deviation of loss in the docstrings.

Cheers,
Wilder

**Comment by wilderrodrigues** [comment]
Hi @tboquet, @titu1994 and @ahundt,

I applied some changes to the code and added initial performance metrics for the MNIST dataset, using a Deep Net.

I'm now running the CNN model and will update the PR later today.

For now, putting it as WIP.

Cheers,
Wilder

**Comment by ahundt** [comment]
thanks, the progress is appreciated!

**Comment by wilderrodrigues** [comment]
Hi again, @tboquet, @titu1994 and @ahundt,

I just pushed some details about IMDb and MNIST performance tests. Could you please have a loo?

Thanks in advance.

Cheers,
Wilder

**Comment by wilderrodrigues** [comment]
Hi @ahundt,

Quick question: do I still need to add the Kaggle Toxicity results? The dataset is quite big and it would take a while to run. I can instead add CIFAR-10, if you think that's okay.

Cheers,
Wilder

**Comment by wilderrodrigues** [comment]
Hi @titu1994, @ahundt and @tboquet,

I know you might be busy, but any extra suggestions  on this?

Thanks again.

Cheers,
Wilder

**Comment by titu1994** [comment]
@wilderrodrigues Sorry for the delay. The updates are appreciated. I think Cifar 10 is a sifficient replacement for Kaggle Toxicity. Cifar 10 is well studied and comparable. However, I don't think there is need for further validation of it's performance. In relation to other advanced activations, they don't provide code or links to dataset validations. 

I think this is ready to merge. Just waiting for @ahundt for a final confirmation

**Comment by wilderrodrigues** [comment]
Thanks @titu1994 and @tboquet!

Will wait for @ahundt's confirmation to start spreading the word about the function. Hopefully, after presenting it the conferences in May, it will become popular and reach Keras - who knows? ;)

If you guys need help with code / reviewing PRs, I can dedicate some of my time to help you out.

Cheers,
Wilder

File: keras_contrib/layers/advanced_activations.py [hunk 4]
```diff
@@ -283,3 +283,103 @@ def get_config(self):
         return dict(list(base_config.items()) + list(config.items()))
 
 get_custom_objects().update({'Swish': Swish})
```
[the remark this answers is not in the document]
Comment (wilderrodrigues) [reply keras_contrib/layers/advanced_activations.py]: Good catch, @ahundt. In my notebooks I needed this when loading the weights of a model that used SineReLU. But then, I did so: ```custom_objects={'SineReLU': SineReLU}``` in the load_model method.

With that line added , I don't need to add the extra parameter. Will push the changes.

Thanks!

**Comment by wilderrodrigues** [comment]
Changes applied and the code works fine, @ahundt.

Thanks again for the catching that one.

Cheers,
Wilder

**Comment by ahundt** [comment]
LGTM, thanks!
